### PR TITLE
feat: Allow context to be overwritten in createTestClient

### PR DESCRIPTION
**Pull request by Pruxis**
In the current implementation you would have to set up a seperate Apollo Server instance with a context function or object that suits the needs of your testing suite.

By allowing to adjust this directly in the createTestClient it takes away this verbosity.

TODO:

* [x] Update CHANGELOG.md with your change (include reference to issue & this PR)
* [x] Make sure all of the significant new logic is covered by tests
* [x] Rebase your changes on master so that they can be merged easily
* [x] Make sure all tests and linter rules pass

### CODE REVIEW

**Comment by Pruxis** [comment]
I have found a bug in the current implementation being that whenever you change the context by calling the createTestClient function you also mutate the context of all following tests that utilise the same server instance.

I'm looking into how I can avoid this behaviour by not mutating the server but injecting the context in a later stage.

**Comment by Pruxis** [comment]
I've updated the implementation to allow to override the context specifically per query / mutation but also by option when creating your test client. If you do not pass this option it's defaulted by the context passed to the server instance.

**Comment by Pruxis** [comment]
It's been over a month, @martijnwalraven @abernix could I atleast get some feedback

File: packages/apollo-server-core/src/ApolloServer.ts [hunk 1]
```diff
@@ -96,8 +96,8 @@ export class ApolloServerBase {
   public subscriptionsPath?: string;
   public graphqlPath: string = '/graphql';
   public requestOptions: Partial<GraphQLOptions<any>> = Object.create(null);
+  public context?: Context | ContextFunction;
```
Comment (abernix) [inline packages/apollo-server-core/src/ApolloServer.ts]: I can see how this is useful functionality to add to the `createTestClient` utility, but I'm afraid that just changing this previously-private `context` of `ApolloServerBase` (which is sub-classed elsewhere) into a public property, free to be manipulated without guard, is over-exposing the `context`.  

Any ideas what another approach here might look like?

**Comment by abernix** [comment]
@Pruxis As an example use case, can you provide a pattern that you're trying to use/avoid? That might help guide the engineering behind this.

For example, are you trying to avoid constructing a new `ApolloServer` for each test which needs a different context?  (I almost think that the accurate construction of that `context` could be beneficial in terms of being explicit and providing a reliable test.).  Alternatively, it's of course possible to declare a `context` function on `ApolloServer` which calls a locally-scoped function within a test:

```
```

**Comment by Pruxis** [comment]
@abernix What I'm trying to accomplish is that I am able to re-use the same ApolloServer configuration / instance that I would normally have when starting up my service.

Leaving me with only one ApolloServer configuration to maintain, manage and test. I have a very heavy permission based application so to test each entry point of different roles, scopes and authorization flags.

I feel like this pattern gives a lot of clarity for those context-dependant applications, every developer can clearly see where the context is overwritten and where it isn't. Instead of having the mental confusion of the mutation context object / function

For instance a simplified version of my tests would look something like this:

```
const createAgent = ({ enableFlag1 } = {}) => ({
  user: {
    name: 'Laurens',
    role: 'AGENT',
    get channels() {
      return [global.__conversations__channelId];
    },
    get skills() {
      return [global.__conversations__skillId];
    },
    get languages() {
      return [global.__conversations__languageId];
    },
    organisation: {
      get _id() {
        // We utilise a getter as the beforeAll hook sets this value.
        return global.__conversations__organisationId;
      },
      settings: { flag1 },
    },
  },
});

const { query } = createTestClient(createAgent());

it('Get conversations if flag one is enabled', async () => {
  const { data: { conversationsForAgent } = {} } = await query({
      query: GET_CONVERSATIONS,
      context: createAgent({ enableSkillsRouting: true }),
    });
});
```

**Comment by Pruxis** [comment]
@abernix Any chance of looking at this to find a solution?

**Comment by Pruxis** [comment]
It's actuality absurd how long PR's keep hanging in apollo-server, I've been using a patch-package with success for the past 3 months due to contributions just being ignored.

I would've loved to find a solution that works and fits the apollo tooling.

**Comment by Pruxis** [comment]
Closed due to inactivity.

I am solving this by executing the following

```
      const server = // get ApolloServer here
      const apolloClient = createTestClient(server);
      global.__apolloTestClient__ = {
        query: ({ query, variables, context = defaultContext || {} }) => {
          server.requestOptions.context = context;
          // @ts-ignore
          return apolloClient.query({ query, variables });
        },
        mutate: ({ mutation, variables, context = defaultContext || {} }) => {
          server.requestOptions.context = context;
          // @ts-ignore
          return apolloClient.mutate({ mutation, variables });
        },
      };
```

**Pull request closed without merging**